elk/utils/multi_gpu.py

```diff
@@ -0,0 +1,184 @@
+        return [self.first_device] + self.other_devices
+
+
+def instantiate_model_with_devices(
+    cfg: "Extract", device_config: ModelDevices, is_verbose: bool, **kwargs
+) -> PreTrainedModel:
+    first_device = device_config.first_device
+    if cfg.int8:
+        # Required by `bitsandbytes`
+        torch_dtype = torch.float16
+    elif device_config == "cpu":
+        torch_dtype = torch.float32
+    else:
+        torch_dtype = "auto"
+
+    # TODO: Maybe we should ensure the device map is the same
+    # for all the extract processes? This is because the device map
+    # can affect performance highly and its annoying if one process
+    # is using a different device map than the others.
+    device_map = (
+        {"": first_device}
+        if device_config.is_single_gpu
+        else create_device_map(
+            model_str=cfg.model,
+            use_8bit=cfg.int8,
+            torch_dtype=torch_dtype,
+            model_devices=device_config,
+            verbose=is_verbose,
+        )
+    )
+    if is_verbose:
+        print(
+            f"Using {len(device_config.used_devices)} gpu(s) to"
+            f" instantiate a single model."
+        )
+    with redirect_stdout(None) if not is_verbose else nullcontext():
+        model = instantiate_model(
+            cfg.model,
+            device_map=device_map,
+            load_in_8bit=cfg.int8,
+            torch_dtype=torch_dtype,
+            **kwargs,
+        )
+    return model
+
+
+def create_device_map(
+    model_str: str,
+    use_8bit: float,
+    torch_dtype: dtype | str,
+    model_devices: ModelDevices,
+    verbose: bool,
+) -> dict[str, str]:
+    """Creates a device map for a model running on multiple GPUs."""
+    with init_empty_weights():
+        # Need to first instantiate an empty model to get the layer class
+        model = instantiate_model(model_str=model_str, torch_dtype=torch_dtype)
+
+    # e.g. {"cuda:0": 16000, "cuda:1": 16000}
+    max_memory_all_devices: dict[str, int] = get_available_memory_for_devices()
+    # now let's get the available memory for the devices we want to use
+    used_devices = model_devices.used_devices
+    max_memory_used_devices: dict[str, int | float] = {
+        device: max_memory_all_devices[device] for device in used_devices
+    }
+    # Decrease the memory potentially used by the first device
+    # because we're going to create additional tensors on it
+    max_memory_used_devices[model_devices.first_device] = (
+        max_memory_used_devices[model_devices.first_device] * 0.6
+    )
+    # If 8bit, multiply the memory by 2
+    # This is because we instantiated our empty model in (probably) float16
+    # We aren't able to instantiate an empty model in 8bit currently
+    max_memory_used_devices = (
+        {
+            device: max_memory_used_devices[device] * 2
+            for device in max_memory_used_devices
+        }
+        if use_8bit
+        else max_memory_used_devices
+    )
```

Comment:
the TLDR is that we need to tell `infer_auto_device_map` that we want to use {"cuda:0": 30gb, "cuda:1": 40gb} and NO OTHER GPUS OR CPU OR DISK